django/db/models/aggregates.py

```diff
@@ -392,3 +394,61 @@ def __init__(self, expression, sample=False, **extra):
 
     def _get_repr_options(self):
         return {**super()._get_repr_options(), "sample": self.function == "VAR_SAMP"}
+
+
+class JSONArrayAgg(Aggregate):
+    function = "JSON_ARRAYAGG"
+    output_field = JSONField()
+    arity = 1
+
+    def as_sql(self, compiler, connection, **extra_context):
+        if self.filter and not connection.features.supports_aggregate_filter_clause:
+            raise NotSupportedError(
+                "JSONArrayAgg(filter) is not supported on this database backend."
+            )
+        return super().as_sql(compiler, connection, **extra_context)
+
+    def as_sqlite(self, compiler, connection, **extra_context):
+        sql, params = self.as_sql(
+            compiler, connection, function="JSON_GROUP_ARRAY", **extra_context
+        )
+        # JSON_GROUP_ARRAY defaults to returning an empty array on an empty set.
+        # Modifies the SQL to support a custom default value to be returned,
+        # if a default argument is not passed, null is returned instead of [].
+        if (default := self.default) == []:
+            return sql, params
+        # Ensure Count() is against the exact same parameters (filter, distinct)
+        count = self.copy()
+        count.__class__ = Count
+        count_sql, count_params = compiler.compile(count)
+        default_sql = ""
+        default_params = ()
+        if default is not None:
+            default_sql, default_params = compiler.compile(default)
+            default_sql = f" ELSE {default_sql}"
+        sql = f"(CASE WHEN {count_sql} > 0 THEN {sql}{default_sql} END)"
+        return sql, count_params + params + default_params
+
+    def as_postgresql(self, compiler, connection, **extra_context):
+        if not connection.features.is_postgresql_16:
+            sql, params = super().as_sql(
+                compiler,
+                connection,
+                function="ARRAY_AGG",
+                **extra_context,
+            )
+            return f"TO_JSONB({sql})", params
+        extra_context.setdefault(
+            "template", "%(function)s(%(distinct)s%(expressions)s RETURNING JSONB)"
+        )
+        return self.as_sql(compiler, connection, **extra_context)
+
+    def as_oracle(self, compiler, connection, **extra_context):
+        # Return same date field format as on other supported backends.
+        expression = self.get_source_expressions()[0]
+        internal_type = expression.output_field.get_internal_type()
+        if internal_type == "DateField":
+            extra_context.setdefault(
+                "template", "%(function)s(TO_CHAR(%(expressions)s, 'YYYY-MM-DD'))"
+            )
```

Comment:
Wow I'm surprised that Oracle doesn't have a proper `DATE` [data type](https://dbfiddle.uk/TzYmP7Kj). I'm not a fan of this logic but I also can't think of a way around it. Maybe it's not a huge deal because `JSONObject` is affected by the same problem and we don't specialize it?

Again the template must include `filter` and `distinct`

```suggestion
        if internal_type == "DateField":
            extra_context.setdefault(
                "template", "%(function)s(%(distinct)sTO_CHAR(%(expressions)s, 'YYYY-MM-DD'))%(filter)s"
            )
```

An alternative way to achieve the same thing without leaning into `template` too much would be to do


```python
expression = self.get_source_expressions()[0]
internal_type = expression.output_field.get_internal_type()
if internal_type == "DateField":
    clone = self.copy()
    clone.set_source_expressions([Func(expression, Value("YYYY-MM-DD"), function="TO_CHAR")])
    return clone.as_sql(compiler, connection, **extra_context)

Reply:
Hey @charettes , thanks for your suggestion  not to use template on datefield for oracle, though am getting the following error 

Traceback (most recent call last):
  File "/home/django/tests/aggregation/tests.py", line 2685, in test_datefield
    vals = Author.objects.aggregate(jsonarrayagg=JSONArrayAgg("book__pubdate"))
  File "/home/django/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/django/django/db/models/query.py", line 589, in aggregate
    return self.query.chain().get_aggregation(self.db, kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/django/django/db/models/sql/query.py", line 626, in get_aggregation
    result = compiler.execute_sql(SINGLE)
  File "/home/django/django/db/models/sql/compiler.py", line 1610, in execute_sql
    sql, params = self.as_sql()
                  ~~~~~~~~~~~^^
  File "/home/django/django/db/models/sql/compiler.py", line 766, in as_sql
    extra_select, order_by, group_by = self.pre_sql_setup(
                                       ~~~~~~~~~~~~~~~~~~^
        with_col_aliases=with_col_aliases or bool(combinator),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/django/django/db/models/sql/compiler.py", line 85, in pre_sql_setup
    self.setup_query(with_col_aliases=with_col_aliases)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/django/db/models/sql/compiler.py", line 74, in setup_query
    self.select, self.klass_info, self.annotation_col_map = self.get_select(
                                                            ~~~~~~~~~~~~~~~^
        with_col_aliases=with_col_aliases,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/django/django/db/models/sql/compiler.py", line 316, in get_select
    sql, params = self.compile(col)
                  ~~~~~~~~~~~~^^^^^
  File "/home/django/django/db/models/sql/compiler.py", line 575, in compile
    sql, params = vendor_impl(self, self.connection)
                  ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/django/db/models/aggregates.py", line 460, in as_oracle
    clone.set_source_expressions(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        [Func(expression, Value("YYYY-MM-DD"), function="TO_CHAR")])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/django/db/models/aggregates.py", line 116, in set_source_expressions
    *exprs, self.filter, self.order_by = exprs
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reply:
You forgot to include the actual exception in here which makes it impossible to diagnose unfortunately.

Reply:
Here is the exception `ValueError: not enough values to unpack (expected at least 2, got 1)`

Reply:
Hey @charettes , any updates or we stick to the template parameter.

Reply:
The code above was meant to point you in the right direction so you can figure it out yourself by moving things around. The full code is

```python
def as_oracle(self, compiler, connection, **extra_context):
        # Oracle turns DATE columns into ISO 8601 timestamp including T00:00:00
        # suffixed when converting to JSON while other backends only include
        # the date part.
        source_expressions = self.get_source_expressions()
        expression = source_expressions[0]
        if isinstance(expression.output_field, DateField):
            clone = self.copy()
            clone.set_source_expressions(
                [
                    Func(expression, Value("YYYY-MM-DD"), function="TO_CHAR"),
                    *source_expressions[1:],
                ]
            )
            return clone.as_sql(compiler, connection, **extra_context)
        return self.as_sql(compiler, connection, **extra_context)
```